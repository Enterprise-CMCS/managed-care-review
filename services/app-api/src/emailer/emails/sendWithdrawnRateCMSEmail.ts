import { formatCalendarDate } from '@mc-review/dates'
import type {
    ProgramType,
    RateRevisionType,
    RateType,
    RateReviewType,
} from '../../domain-models'
import type {
    EmailData,
    StateAnalystsEmails,
    EmailConfiguration,
} from '../emailer'
import {
    findContractPrograms,
    renderTemplate,
    stripHTMLFromTemplate,
} from '../templateHelpers'
import { packageName as generatePackageName } from '@mc-review/submissions'
import { rateSummaryURL, submissionSummaryURL } from '../generateURLs'
import { pruneDuplicateEmails } from '../formatters'

type WithdrawnFromContractData = {
    contractName: string
    submissionURL: string
}

type WithdrawnRateEtaData = {
    rateName: string
    withdrawnBy: string //email
    withdrawnDate: string // mm/dd/yyyy format in PT timezone.
    withdrawnReason: string
    withdrawnFromContractData: WithdrawnFromContractData[]
    summaryURL: string // rate summary url
}

type ValidatedWithdrawnRate = {
    latestRateRev: RateRevisionType
    latestAction: RateReviewType
    withdrawnFromContractData: WithdrawnFromContractData[]
    stateContactEmails: string[]
}

/**
 * Validates and parses a withdrawn rate, ensuring it meets specific criteria and extracting relevant data.
 *
 * @param {RateType} rate - The rate object to validate and parse.
 * @param {ProgramType[]} statePrograms - An array of state programs to be used for validation.
 * @param {EmailConfiguration} config - Configuration object containing email-related settings.
 * @returns {Error | ValidatedWithdrawnRate} - Returns an Error object if validation fails, otherwise returns a ValidatedWithdrawnRate object containing parsed data.
 **/
export const validateAndParseWithdrawnRate = (
    rate: RateType,
    statePrograms: ProgramType[],
    config: EmailConfiguration
): Error | ValidatedWithdrawnRate => {
    if (rate.consolidatedStatus !== 'WITHDRAWN') {
        return new Error('Rate consolidated status is not WITHDRAWN')
    }

    //check to make sure rate has a withdrawn action
    const reviewStatusActions = rate.reviewStatusActions

    if (!reviewStatusActions || reviewStatusActions.length === 0) {
        return new Error('Rate does not have any review actions')
    }

    const latestAction = reviewStatusActions.sort(
        (a, b) =>
            new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
    )[0]

    if (latestAction.actionType !== 'WITHDRAW') {
        return new Error('Rate latest review action was not a withdraw action')
    }

    // These contracts will not include linked contracts that are unlocked or in draft
    const withdrawnFromContracts = rate.withdrawnFromContracts ?? []

    const latestRateRev = rate.packageSubmissions[0].rateRevision
    if (!latestRateRev) {
        return new Error('Rate does not have a rate revision')
    }

    const withdrawnFromContractData: WithdrawnFromContractData[] = []
    const stateContactEmails: string[] = []

    if (withdrawnFromContracts.length > 0) {
        for (const contract of withdrawnFromContracts) {
            // Get latest revision to generate package name. Use draft revision if contract is unlocked.
            const latestContractRev =
                contract.consolidatedStatus === 'UNLOCKED'
                    ? contract.draftRevision!
                    : contract.packageSubmissions[0].contractRevision

            const pkgPrograms = findContractPrograms(
                latestContractRev,
                statePrograms
            )

            if (pkgPrograms instanceof Error) {
                return new Error(
                    `Error parsing withdrawn from contract data for contract with ID: ${contract.id}. ${pkgPrograms.message}`
                )
            }

            const packageName = generatePackageName(
                contract.stateCode,
                contract.stateNumber,
                latestContractRev.formData.programIDs,
                pkgPrograms
            )

            const submissionURL = submissionSummaryURL(
                contract.id,
                config.baseUrl
            )

            withdrawnFromContractData.push({
                contractName: packageName,
                submissionURL,
            })

            latestContractRev.formData.stateContacts.forEach((contact) => {
                if (contact.email) stateContactEmails.push(contact.email)
            })
        }
    } else {
        // If rate was orphaned rate, then we use the last rate submission that had contracts to send the email to.
        const latestSubmissionWithContracts = rate.packageSubmissions.find(
            (pkg) => pkg.contractRevisions.length > 0
        )

        if (latestSubmissionWithContracts) {
            for (const cr of latestSubmissionWithContracts.contractRevisions) {
                cr.formData.stateContacts.forEach((contact) => {
                    if (contact.email) stateContactEmails.push(contact.email)
                })
            }
        }
    }

    return {
        latestRateRev,
        latestAction,
        withdrawnFromContractData,
        stateContactEmails,
    }
}

export const sendWithdrawnRateCMSEmail = async (
    config: EmailConfiguration,
    rate: RateType,
    statePrograms: ProgramType[],
    stateAnalystsEmails: StateAnalystsEmails
): Promise<EmailData | Error> => {
    // validate the withdrawn rate
    const withdrawnRateData = validateAndParseWithdrawnRate(
        rate,
        statePrograms,
        config
    )

    if (withdrawnRateData instanceof Error) {
        return withdrawnRateData
    }

    const { latestRateRev, latestAction, withdrawnFromContractData } =
        withdrawnRateData

    const toAddresses = pruneDuplicateEmails([
        ...stateAnalystsEmails,
        ...config.dmcpSubmissionEmails,
        ...config.oactEmails,
        ...config.devReviewTeamEmails,
    ])

    const summaryURL = rateSummaryURL(rate.id, config.baseUrl)

    const etaData: WithdrawnRateEtaData = {
        rateName: latestRateRev.formData.rateCertificationName!,
        withdrawnBy: latestAction.updatedBy.email,
        withdrawnDate: formatCalendarDate(
            latestAction.updatedAt,
            'America/Los_Angeles'
        ),
        withdrawnReason: latestAction.updatedReason,
        withdrawnFromContractData,
        summaryURL,
    }

    const template = await renderTemplate<WithdrawnRateEtaData>(
        'sendWithdrawnRateCMSEmail',
        etaData
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses,
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }${etaData.rateName} was withdrawn`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
