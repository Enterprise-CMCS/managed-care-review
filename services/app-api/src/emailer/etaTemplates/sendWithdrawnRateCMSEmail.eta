<h1 style=<%~'"font-weight: normal;"'%>>Rate <%= it.rateName %> was withdrawn by CMS</h1>
<b>Withdrawn by:</b> <%= it.withdrawnBy %><br />
<b>Withdrawn on:</b> <%= it.withdrawnDate %><br />
<b>Reason for withdrawal:</b> <%= it.withdrawnReason %><br />
<% if (it.withdrawnFromContractData.length > 1) { %>
<b>Withdrawn from:</b>
<ul>
<% it.withdrawnFromContractData.forEach(function(contract){ %>
    <li><%= contract.contractName %></li>
<% }) %>
</ul>
<% } else if (it.withdrawnFromContractData[0]) { %>
<b>Withdrawn from:</b>
<%= it.withdrawnFromContractData[0].contractName %><br />
<br />
<% } %>
<a href="<%= it.summaryURL %>">View withdrawn rate in MC-Review</a>
