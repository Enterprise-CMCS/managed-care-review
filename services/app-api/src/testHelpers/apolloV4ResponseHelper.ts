import type { GraphQLFormattedError } from 'graphql'
import type { ApolloServer } from '@apollo/server'

export type TestGraphQLResponse<T = any> = {
    data?: T
    errors?: GraphQLFormattedError[]
}

export function extractGraphQLResponse<T = any>(
    response: any
): TestGraphQLResponse<T> {
    if ('body' in response && response.body) {
        const result =
            response.body.kind === 'single'
                ? response.body.singleResult
                : response.body

        return {
            data: result.data,
            errors: result.errors,
        }
    }

    return {
        data: response.data,
        errors: response.errors,
    }
}

export async function executeGraphQLOperation<TData = any, TVariables = any>(
    server: ApolloServer,
    options: {
        query: any
        variables?: TVariables
        contextValue?: any
    }
): Promise<TestGraphQLResponse<TData>> {
    const response = await server.executeOperation({
        query: options.query,
        variables: options.variables as any,
    })

    return extractGraphQLResponse<TData>(response)
}
