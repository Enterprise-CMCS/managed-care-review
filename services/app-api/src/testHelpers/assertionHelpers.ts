// Functions that assists with type narrow in jest tests

// must -  interrupts test flow to through an error
function must<T>(maybeErr: T | Error): T {
    if (maybeErr instanceof Error) {
        throw maybeErr
    }
    return maybeErr
}

// expectToBeDefined - properly  type narrows the results of toBeDefined
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41179 would eventually address this but is not implemented
function expectToBeDefined<T>(arg: T): asserts arg is NonNullable<T> {
    expect(arg).toBeDefined()
}

export { must, expectToBeDefined }
