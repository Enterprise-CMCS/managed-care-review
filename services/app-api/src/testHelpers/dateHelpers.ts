// returns today's date in the ISO standard form. like '2022-02-24'
function todaysDate(): string {
    const today = new Date()
    return today.toISOString().split('T')[0]
}

export { todaysDate }
