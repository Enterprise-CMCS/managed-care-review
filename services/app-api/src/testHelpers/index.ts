export {
    assertAnError,
    assertAnErrorExtensions,
    assertAnErrorCode,
} from './gqlAssertions'

export { must } from './assertionHelpers'

export {
    mockInsertContractArgs,
    mockContractRevision,
} from './contractDataMocks'

export {
    mockInsertRateArgs,
    mockRateRevision,
    mockDraftRate,
} from './rateDataMocks'

export { testS3Client } from './s3Helpers'
export { getStateRecord } from './stateHelpers'

export { consoleLogFullData } from './debugHelpers'

export {
    fetchTestRateById,
    submitTestRate,
    unlockTestRate,
    updateTestRate,
} from './gqlRateHelpers'

export {
    createAndSubmitTestContract,
    fetchTestContract,
    approveTestContract,
    updateTestContractDraftRevision,
    createTestContract,
    fetchTestContractWithQuestions,
} from './gqlContractHelpers'

export {
    clearDocMetadata,
    clearMetadataFromContractFormData,
    clearMetadataFromRateFormData,
} from './documentHelpers'

export { mockGqlContractDraftRevisionFormDataInput } from './gqlContractInputMocks'
