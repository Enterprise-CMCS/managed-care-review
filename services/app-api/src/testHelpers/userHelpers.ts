import { NewPostgresStore, type InsertUserArgsType } from '../postgres'
import type {
    AdminUserType,
    CMSApproverUserType,
    CMSUserType,
    StateUserType,
    UserType,
    BusinessOwnerUserType,
    HelpdeskUserType,
} from '../domain-models'
import { sharedTestPrismaClient } from './storeHelpers'
import { v4 as uuidv4 } from 'uuid'

const testCMSUser = (userData?: Partial<CMSUserType>): CMSUserType => ({
    id: uuidv4(),
    role: 'CMS_USER',
    email: 'zuko@example.com',
    familyName: 'Zuko',
    givenName: 'Prince',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testCMSApproverUser = (
    userData?: Partial<CMSApproverUserType>
): CMSApproverUserType => ({
    id: uuidv4(),
    role: 'CMS_APPROVER_USER',
    email: 'azula@example.com',
    familyName: 'Azula',
    givenName: 'Princess',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testStateUser = (userData?: Partial<StateUserType>): StateUserType => ({
    id: uuidv4(),
    stateCode: 'FL',
    role: 'STATE_USER',
    email: 'james@example.com',
    familyName: 'Brown',
    givenName: 'James',
    ...userData,
})

const testAdminUser = (userData?: Partial<AdminUserType>): AdminUserType => ({
    id: uuidv4(),
    role: 'ADMIN_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const testBusinessOwnerUser = (
    userData?: Partial<BusinessOwnerUserType>
): BusinessOwnerUserType => ({
    id: uuidv4(),
    role: 'BUSINESSOWNER_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const testHelpdeskUser = (
    userData?: Partial<HelpdeskUserType>
): HelpdeskUserType => ({
    id: uuidv4(),
    role: 'HELPDESK_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const createDBUsersWithFullData = async (
    users: UserType[]
): Promise<UserType[]> => {
    const prismaClient = await sharedTestPrismaClient()
    const postgresStore = NewPostgresStore(prismaClient)

    const usersSeed: InsertUserArgsType[] = users.map((user: UserType) => {
        const userArgs: InsertUserArgsType & { id?: string } = {
            ...user,
            userID: user.id,
        }

        delete userArgs.id

        return userArgs as InsertUserArgsType
    })

    const result = await postgresStore.insertManyUsers(usersSeed)

    if (result instanceof Error) {
        throw result
    }

    return result
}

const iterableCmsUsersMockData: {
    userRole: 'CMS_USER' | 'CMS_APPROVER_USER'
    mockUser: <T>(userData?: Partial<T>) => CMSUserType | CMSApproverUserType
}[] = [
    {
        userRole: 'CMS_USER',
        mockUser: testCMSUser,
    },
    {
        userRole: 'CMS_APPROVER_USER',
        mockUser: testCMSApproverUser,
    },
]

const iterableNonCMSUsersMockData: {
    userRole:
        | 'HELPDESK_USER'
        | 'BUSINESSOWNER_USER'
        | 'ADMIN_USER'
        | 'STATE_USER'
    mockUser: <T>(
        userData?: Partial<T>
    ) =>
        | AdminUserType
        | BusinessOwnerUserType
        | HelpdeskUserType
        | StateUserType
}[] = [
    {
        userRole: 'ADMIN_USER',
        mockUser: testAdminUser,
    },
    {
        userRole: 'BUSINESSOWNER_USER',
        mockUser: testBusinessOwnerUser,
    },
    {
        userRole: 'HELPDESK_USER',
        mockUser: testHelpdeskUser,
    },
    {
        userRole: 'STATE_USER',
        mockUser: testStateUser,
    },
]

export {
    testAdminUser,
    testStateUser,
    testCMSUser,
    createDBUsersWithFullData,
    iterableCmsUsersMockData,
    iterableNonCMSUsersMockData,
    testCMSApproverUser,
    testBusinessOwnerUser,
    testHelpdeskUser,
}
