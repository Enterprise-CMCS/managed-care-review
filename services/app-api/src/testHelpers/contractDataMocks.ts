import { v4 as uuidv4 } from 'uuid'
import type {
    ContractRevisionTableWithRates,
    ContractTableFullPayload,
} from '../postgres/contractAndRates/prismaFullContractRateHelpers'
import type { StateCodeType } from '@mc-review/submissions'
import type { ContractFormDataType } from '../domain-models'
import { findStatePrograms, type InsertContractArgsType } from '../postgres'
import { must } from './assertionHelpers'

const defaultContractData = () => ({
    id: uuidv4(),
    createdAt: new Date(),
    updatedAt: new Date(),
    mccrsID: null,
    stateCode: 'MN',
    stateNumber: 111,
})

const mockInsertContractArgs = ({
    stateCode = 'MN',
    ...formData
}: {
    stateCode?: StateCodeType
} & Partial<ContractFormDataType>): InsertContractArgsType => {
    const statePrograms = must(findStatePrograms(stateCode))

    return {
        stateCode: stateCode,
        submissionType: formData?.submissionType ?? 'CONTRACT_AND_RATES',
        submissionDescription:
            formData?.submissionDescription ?? 'Contract 1.0',
        contractType: formData?.contractType ?? 'BASE',
        programIDs: formData?.programIDs ?? [statePrograms[0].id],
        populationCovered: formData?.populationCovered ?? 'MEDICAID',
        riskBasedContract: formData?.riskBasedContract ?? false,
        ...formData,
    }
}

const mockContractRevision = (
    contract?: Partial<ContractTableFullPayload>,
    revision?: Partial<ContractRevisionTableWithRates>,
    stateCode: StateCodeType = 'MN'
): ContractRevisionTableWithRates => {
    const statePrograms = must(findStatePrograms(stateCode))

    return {
        id: uuidv4(),
        contract: {
            ...defaultContractData(),
            ...contract,
        },
        relatedSubmisions: [],
        createdAt: new Date(),
        updatedAt: new Date(),
        submitInfo: {
            id: uuidv4(),
            updatedAt: new Date(),
            updatedByID: 'someone',
            updatedReason: 'submit',
            updatedBy: {
                id: 'someone',
                createdAt: new Date(),
                updatedAt: new Date(),
                givenName: 'Bob',
                familyName: 'Law',
                email: 'boblaw@example.com',
                role: 'STATE_USER',
                divisionAssignment: null,
                stateCode: stateCode,
            },
        },
        dsnpContract: null,
        unlockInfo: null,
        contractID: 'contractID',
        submitInfoID: null,
        unlockInfoID: null,
        programIDs: [statePrograms[0].id],
        populationCovered: 'MEDICAID' as const,
        submissionType: 'CONTRACT_ONLY' as const,
        riskBasedContract: false,
        submissionDescription: 'Test',
        stateContacts: [],
        supportingDocuments: [
            {
                id: uuidv4(),
                position: 0,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract supporting doc',
                s3URL: 's3://bucketname/key/test1',
                sha256: '2342fwlkdmwvw',
                dateAdded: new Date(),
            },
            {
                id: uuidv4(),
                position: 1,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract supporting doc 2',
                s3URL: 's3://bucketname/key/test1',
                sha256: '45662342fwlkdmwvw',
                dateAdded: new Date(),
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                id: uuidv4(),
                position: 0,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract doc',
                s3URL: 's3://bucketname/key/test1',
                sha256: '8984234fwlkdmwvw',
                dateAdded: new Date(),
            },
        ],
        contractDateStart: new Date(Date.UTC(2025, 5, 1)),
        contractDateEnd: new Date(Date.UTC(2026, 4, 30)),
        managedCareEntities: ['MCO'],
        federalAuthorities: ['STATE_PLAN' as const],
        modifiedBenefitsProvided: false,
        modifiedGeoAreaServed: false,
        modifiedMedicaidBeneficiaries: false,
        modifiedRiskSharingStrategy: false,
        modifiedIncentiveArrangements: false,
        modifiedWitholdAgreements: false,
        modifiedStateDirectedPayments: false,
        modifiedPassThroughPayments: false,
        modifiedPaymentsForMentalDiseaseInstitutions: false,
        modifiedMedicalLossRatioStandards: false,
        modifiedOtherFinancialPaymentIncentive: false,
        modifiedEnrollmentProcess: false,
        modifiedGrevienceAndAppeal: false,
        modifiedNetworkAdequacyStandards: true,
        modifiedLengthOfContract: true,
        modifiedNonRiskPaymentArrangements: null,
        inLieuServicesAndSettings: null,
        statutoryRegulatoryAttestation: null,
        statutoryRegulatoryAttestationDescription: null,
        ...revision,
    }
}

export { mockInsertContractArgs, mockContractRevision }
