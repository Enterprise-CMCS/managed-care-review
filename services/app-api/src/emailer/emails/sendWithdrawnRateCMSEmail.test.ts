import {
    mockContract,
    mockContractRev,
    mockMNState,
    mockRate,
    testEmailConfig,
    testStateAnalystsEmails,
} from '../../testHelpers/emailerHelpers'
import { testCMSUser, testStateUser } from '../../testHelpers/userHelpers'
import { sendWithdrawnRateCMSEmail } from './sendWithdrawnRateCMSEmail'

const testRate = mockRate({
    id: 'test-rate',
    createdAt: new Date('2024-04-12'),
    updatedAt: new Date('2024-04-12'),
    status: 'RESUBMITTED',
    reviewStatus: 'WITHDRAWN',
    consolidatedStatus: 'WITHDRAWN',
    parentContractID: 'parent-contract',
    withdrawnFromContracts: [
        {
            ...mockContract({
                id: 'parent-contract',
                stateNumber: 100,
                packageSubmissions: [
                    {
                        submitInfo: {
                            updatedAt: new Date('2024-04-12'),
                            updatedReason: 'CMS withdrew rate',
                            updatedBy: testStateUser({
                                email: 'cms@cms.com',
                            }),
                        },
                        submittedRevisions: [],
                        rateRevisions: [],
                        contractRevision: mockContractRev({
                            submitInfo: {
                                updatedAt: new Date('2024-04-12'),
                                updatedReason: 'CMS withdrew rate',
                                updatedBy: testCMSUser({
                                    email: 'cms@cms.com',
                                }),
                            },
                            formData: {
                                ...mockContractRev().formData,
                                stateContacts: [
                                    {
                                        name: 'parent-contract-state-contact',
                                        titleRole: 'state contact',
                                        email: 'parent-contract-state-contact@state.com',
                                    },
                                ],
                            },
                        }),
                    },
                ],
            }),
        },
        {
            ...mockContract({
                id: 'linked-contract',
                stateNumber: 222,
                packageSubmissions: [
                    {
                        submitInfo: {
                            updatedAt: new Date('2024-04-12'),
                            updatedReason: 'CMS withdrew rate',
                            updatedBy: testCMSUser({
                                email: 'cms@cms.com',
                            }),
                        },
                        submittedRevisions: [],
                        rateRevisions: [],
                        contractRevision: mockContractRev({
                            submitInfo: {
                                updatedAt: new Date('2024-04-12'),
                                updatedReason: 'CMS withdrew rate',
                                updatedBy: testCMSUser({
                                    email: 'cms@cms.com',
                                }),
                            },
                            formData: {
                                ...mockContractRev().formData,
                                stateContacts: [
                                    {
                                        name: 'linked-contract-state-contact',
                                        titleRole: 'state contact',
                                        email: 'linked-contract-state-contact@state.com',
                                    },
                                    {
                                        name: 'duplicate-state-contact',
                                        titleRole: 'state contact',
                                        email: 'duplicateContact@state-contact.com',
                                    },
                                    {
                                        name: 'duplicate-state-contact',
                                        titleRole: 'state contact',
                                        email: 'duplicateContact@state-contact.com',
                                    },
                                ],
                            },
                        }),
                    },
                ],
            }),
        },
    ],
    packageSubmissions: [
        {
            submitInfo: {
                updatedAt: new Date('2024-04-12'),
                updatedReason: 'CMS withdrawn this rate',
                updatedBy: testStateUser({
                    email: 'parent-contract-submitter@state.com',
                }),
            },
            submittedRevisions: [],
            rateRevision: {
                id: 'test-rate-revision',
                rateID: 'test-rate',
                submitInfo: {
                    updatedAt: new Date('2024-04-12'),
                    updatedReason: 'CMS withdrawn this rate',
                    updatedBy: testStateUser({
                        email: 'parent-contract-submitter@state.com',
                    }),
                },
                createdAt: new Date('2024-04-12'),
                updatedAt: new Date('2024-04-12'),
                formData: {
                    rateType: 'NEW',
                    rateCapitationType: 'RATE_CELL',
                    rateDateStart: new Date('2024-01-01'),
                    rateDateEnd: new Date('2025-01-01'),
                    rateDateCertified: new Date('2024-01-02'),
                    rateCertificationName: 'test-rate-certification-name',
                    rateProgramIDs: [mockMNState().programs[0].id],
                    rateDocuments: [
                        {
                            s3URL: 's3://bucketname/key/test1',
                            name: 'ratedoc1.doc',
                            sha256: 'foobar',
                        },
                    ],
                    supportingDocuments: [],
                    certifyingActuaryContacts: [
                        {
                            name: 'Foo Person',
                            titleRole: 'Bar Job',
                            email: 'certifyingActuary@example.com',
                            actuarialFirm: 'GUIDEHOUSE',
                        },
                    ],
                    addtlActuaryContacts: [
                        {
                            name: 'Bar Person',
                            titleRole: 'Baz Job',
                            email: 'addtlActuaryContacts@example.com',
                            actuarialFirm: 'OTHER',
                            actuarialFirmOther: 'Some Firm',
                        },
                    ],
                    actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                },
            },
            contractRevisions: [],
        },
    ],
    reviewStatusActions: [
        {
            rateID: 'test-rate',
            updatedBy: testCMSUser(),
            updatedReason: 'Test withdraw',
            actionType: 'WITHDRAW',
            updatedAt: new Date('2024-04-12'),
        },
    ],
})

const statePrograms = mockMNState().programs

const stateAnalystsEmails = () => [...testStateAnalystsEmails]

describe('sendWithdrawnRateStateEmail', () => {
    it('includes expected CMS contacts in toAddress', async () => {
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            testRate,
            statePrograms,
            stateAnalystsEmails()
        )

        if (template instanceof Error) {
            throw template
        }

        // Expect state analysts, DMCP submission emails, and OACT emails
        expect(template.toAddresses).toEqual(
            expect.arrayContaining([
                ...stateAnalystsEmails(),
                ...testEmailConfig().dmcpSubmissionEmails,
                ...testEmailConfig().oactEmails,
            ])
        )
    })

    it('renders withdrawn rate with single withdrawn from contract as expected', async () => {
        const singleContractRate = {
            ...testRate,
            withdrawnFromContracts: [
                {
                    ...mockContract({
                        id: 'parent-contract',
                        stateNumber: 100,
                        packageSubmissions: [
                            {
                                submitInfo: {
                                    updatedAt: new Date('2024-04-12'),
                                    updatedReason: 'CMS withdrew rate',
                                    updatedBy: testStateUser({
                                        email: 'cms@cms.com',
                                    }),
                                },
                                submittedRevisions: [],
                                rateRevisions: [],
                                contractRevision: mockContractRev({
                                    submitInfo: {
                                        updatedAt: new Date('2024-04-12'),
                                        updatedReason: 'CMS withdrew rate',
                                        updatedBy: testCMSUser({
                                            email: 'cms@cms.com',
                                        }),
                                    },
                                    formData: {
                                        ...mockContractRev().formData,
                                        stateContacts: [
                                            {
                                                name: 'parent-contract-state-contact',
                                                titleRole: 'state contact',
                                                email: 'parent-contract-state-contact@state.com',
                                            },
                                        ],
                                    },
                                }),
                            },
                        ],
                    }),
                },
            ],
        }
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            singleContractRate,
            statePrograms,
            stateAnalystsEmails()
        )

        if (template instanceof Error) {
            throw template
        }

        // does not contain a ul element
        expect(template.bodyHTML).not.toEqual(expect.stringContaining('<ul>'))

        // expect single contract name
        expect(template.bodyHTML).toEqual(
            expect.stringContaining('MCR-MN-0100-SNBC')
        )

        // expect View the submission link
        expect(template.bodyHTML).toEqual(
            expect.stringContaining(
                '<a href="http://localhost/rates/test-rate">View withdrawn rate in MC-Review</a>'
            )
        )
    })

    it('renders overall email for a withdrawn rate as expected', async () => {
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            testRate,
            statePrograms,
            stateAnalystsEmails()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.bodyHTML).toMatchSnapshot()
    })
})

describe('sendWithdrawnRateStateEmail error handling', () => {
    it('returns an error if rate consolidatedStatus is not WITHDRAWN', async () => {
        const rateWithBadStatus = {
            ...testRate,
            consolidatedStatus: 'SUBMITTED' as const,
        }
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            rateWithBadStatus,
            statePrograms,
            stateAnalystsEmails()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate consolidated status is not WITHDRAWN'
        )
    })
    it('returns an error if no reviewStatusActions are present', async () => {
        const rateWithoutActions = {
            ...testRate,
            reviewStatusActions: [],
        }
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            rateWithoutActions,
            statePrograms,
            stateAnalystsEmails()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate does not have any review actions'
        )
    })
    it('returns an error if latest reviewStatusActions action was not withdraw action', async () => {
        const rateWithoutActions = {
            ...testRate,
            reviewStatusActions: [
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test under review',
                    actionType: 'UNDER_REVIEW' as const,
                    updatedAt: new Date('2024-04-11'),
                },
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test under review',
                    actionType: 'UNDER_REVIEW' as const,
                    updatedAt: new Date('2024-04-15'),
                },
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test withdraw',
                    actionType: 'WITHDRAW' as const,
                    updatedAt: new Date('2024-04-12'),
                },
            ],
        }
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            rateWithoutActions,
            statePrograms,
            stateAnalystsEmails()
        )
        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate latest review action was not a withdraw action'
        )
    })
    it('returns an error when parsing withdrawn from contract data fails', async () => {
        const rateWithBadContractData = {
            ...testRate,
        }

        if (!rateWithBadContractData.withdrawnFromContracts?.[0]) {
            throw new Error('Expected rate to have withdrawn from contracts')
        }

        rateWithBadContractData.withdrawnFromContracts[0].packageSubmissions[0].contractRevision.formData.programIDs =
            ['bad-program-id']

        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            rateWithBadContractData,
            statePrograms,
            stateAnalystsEmails()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            "Error parsing withdrawn from contract data for contract with ID: parent-contract. Can't find programs bad-program-id from state MN"
        )
    })
    it('returns an error when rate has no withdrawn from contracts', async () => {
        const rateWithoutContracts = {
            ...testRate,
            withdrawnFromContracts: [],
        }
        const template = await sendWithdrawnRateCMSEmail(
            testEmailConfig(),
            rateWithoutContracts,
            statePrograms,
            stateAnalystsEmails()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate was withdrawn, but was not associated with any contracts'
        )
    })
})
