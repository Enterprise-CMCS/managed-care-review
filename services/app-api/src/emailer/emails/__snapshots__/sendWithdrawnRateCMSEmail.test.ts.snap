// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendWithdrawnRateCMSEmail > renders overall email for a withdrawn rate as expected 1`] = `
"<h1 style="font-weight: normal;">Rate test-rate-certification-name was withdrawn by CMS</h1>
<b>Withdrawn by:</b> zuko@example.com<br />
<b>Withdrawn on:</b> 04/11/2024<br />
<b>Reason for withdrawal:</b> Test withdraw<br />
<b>Withdrawn from:</b>
<ul>
    <li>MCR-MN-0100-SNBC</li>
    <li>MCR-MN-0222-SNBC</li>
</ul>
<a href="http://localhost/rates/test-rate">View withdrawn rate in MC-Review</a>
"
`;
